### PR TITLE
Add test case for BasePresenter

### PR DESCRIPTION
**Pull request by russhwolf**

### CODE REVIEW

**Comment by ghost** [comment]
[55% (+55.37%) vs master 0%](https://ci.intrepid.io/job/Android/job/pull-requests/job/skeleton-android-pr/91/)

File: app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import io.reactivex.disposables.Disposable;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class BasePresenterTest extends PresenterTestBase<BasePresenter<BaseContract.View>> {
```
Comment (gdaniels) [inline app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java]: No real comment here, but that is quite a line of source code there. :)

**Review by gdaniels — COMMENTED** [review]
I like this better with `Mockito.spy()` -

```
package io.intrepid.skeleton.base;

import android.support.annotation.NonNull;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;

import io.intrepid.skeleton.testutils.PresenterTestBase;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class BasePresenterTest extends PresenterTestBase<BasePresenter<BaseContract.View>> {
    @Mock
    private BaseContract.View view;

    private TestPresenter spyPresenter;

    private Disposable disposable;

    @Before
    public void setUp() {
        spyPresenter = Mockito.spy(new TestPresenter(view, testConfiguration));
        presenter = spyPresenter;
    }

    @Test
    public void bind_onlyOnce() throws Exception {
        presenter.bindView(view);
        presenter.bindView(view);
        assertEquals(view, presenter.view);
        verify(spyPresenter, times(1)).onViewBound();
    }

    @Test
    public void unbind_onlyOnce() throws Exception {
        presenter.unbindView();
        verify(spyPresenter, never()).onViewUnbound();

        presenter.bindView(view);
        presenter.unbindView();
        presenter.unbindView();
        verify(spyPresenter, times(1)).onViewUnbound();
    }

    @Test
    public void unbind_clearsDisposables() {
        presenter.bindView(view);
        assertFalse(disposable.isDisposed());

        presenter.unbindView();
        assertTrue(disposable.isDisposed());
    }

    private class TestPresenter extends BasePresenter<BaseContract.View> {

        TestPresenter(@NonNull BaseContract.View view,
                      @NonNull PresenterConfiguration configuration) {
            super(view, configuration);
        }

        @Override
        protected void onViewBound() {
            disposable = Observable.never().subscribe();
            disposables.add(disposable);
        }
    }

}
```

WDYT?

File: app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        presenter.bindView(view);
+        presenter.bindView(view);
+        assertEquals(view, presenter.view);
+        verify(onBind, times(1)).run();
+    }
+
+    @Test
+    public void unbind_onlyOnce() throws Exception {
+        presenter.unbindView();
+        verify(onUnbind, never()).run();
```
Comment (gdaniels) [inline app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java]: Looking at this again, I'd actually split this into two tests, `presenterWithoutViewDoesNotCallOnViewUnbound()` (or whatever) and `unbind_onlyOnce()`.

File: app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java (hunk 1 above)
Comment (russhwolf) [reply app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java]: Naming is hard 😄

File: app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java (hunk 2 above)
Comment (russhwolf) [reply app/src/test/java/io/intrepid/skeleton/base/BasePresenterTest.java]: Yeah that's a good call. I was bouncing around a bit in how I split these up since there's a bit of duplication due to lifecycle necessity, but this can definitely be split in two.

**Comment by russhwolf** [comment]
To recap some in-person conversation RE Glen's point above, I recently had some issues on Motiv (cc @streetsofboston) where we were having issues with breakpoints in our tests when using `spy()`, so I've been wary of it recently (it's also somewhat discouraged in the Mockito documentation). Looking back at it, it seems that was specifically due to spying on anonymous classes which isn't what's happening here, so I'll do that refactor.

**Comment by ghost** [comment]
[56% (+56.2%) vs master 0%](https://ci.intrepid.io/job/Android/job/pull-requests/job/skeleton-android-pr/92/)

**Comment by russhwolf** [comment]
Heh, using spy() upped the coverage by one line, since we call into the real `BasePresenter.onViewUnbound()` instead of overriding it in `TestPresenter`. Should I increase it further by adding no-op tests that call into the rest of the empty default `BasePresenter` methods? They're all empty so there's nothing to test except that it doesn't crash.

**Comment by russhwolf** [comment]
@gdaniels This has been sitting for a while. Any more thoughts?

**Review by prfarlow1 — APPROVED** [review]
